From: James McCoy <jamessan@debian.org>
Date: Tue, 6 Oct 2015 23:46:30 -0400
Subject: Support sourcing a vimrc.tiny when Vim is invoked as vi

This is used only in the vim-tiny package to allow a specific
configuration for vim-tiny's vi.  The vim-tiny package is substantially
different from other Vim packages, so it does not make sense to
share the same config.

Closes: #222138
Signed-off-by: Stefano Zacchiroli <zack@debian.org>
Signed-off-by: James Vega <jamessan@debian.org>
---
 src/main.c    | 14 +++++++++++++-
 src/os_unix.h |  3 +++
 src/structs.h |  3 +++
 3 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/src/main.c b/src/main.c
index df204bc..cb31bf1 100644
--- a/src/main.c
+++ b/src/main.c
@@ -1786,6 +1786,10 @@ parse_command_name(mparm_T *parmp)
     }
     else if (STRNICMP(initstr, "vim", 3) == 0)
 	initstr += 3;
+#ifdef SYS_TINYRC_FILE
+    else if (STRNICMP(initstr, "vi", 2) == 0)
+	parmp->vi_mode = TRUE;
+#endif
 
     // Catch "[r][g]vimdiff" and "[r][g]viewdiff".
     if (STRICMP(initstr, "diff") == 0)
@@ -3032,7 +3036,12 @@ source_startup_scripts(mparm_T *parmp)
 	 * Get system wide defaults, if the file name is defined.
 	 */
 #ifdef SYS_VIMRC_FILE
-	(void)do_source((char_u *)SYS_VIMRC_FILE, FALSE, DOSO_NONE);
+# if defined(SYS_TINYRC_FILE) && defined(TINY_VIMRC)
+	if (parmp->vi_mode)
+	    (void)do_source((char_u *)SYS_TINYRC_FILE, FALSE, DOSO_NONE);
+	else
+# endif
+	    (void)do_source((char_u *)SYS_VIMRC_FILE, FALSE, DOSO_NONE);
 #endif
 #ifdef MACOS_X
 	(void)do_source((char_u *)"$VIMRUNTIME/macmap.vim", FALSE, DOSO_NONE);
@@ -3067,6 +3076,9 @@ source_startup_scripts(mparm_T *parmp)
 		&& do_source((char_u *)USR_EXRC_FILE, FALSE, DOSO_NONE) == FAIL
 #ifdef USR_EXRC_FILE2
 		&& do_source((char_u *)USR_EXRC_FILE2, FALSE, DOSO_NONE) == FAIL
+#endif
+#if defined(SYS_TINYRC_FILE) && defined(TINY_VIMRC)
+		&& !parmp->vi_mode
 #endif
 		&& !has_dash_c_arg)
 	    {
diff --git a/src/os_unix.h b/src/os_unix.h
index 8919ff0..fcc82eb 100644
--- a/src/os_unix.h
+++ b/src/os_unix.h
@@ -213,6 +213,9 @@ typedef struct dsc$descriptor   DESC;
 /*
  * Unix system-dependent file names
  */
+#ifndef SYS_TINYRC_FILE
+# define SYS_TINYRC_FILE "$VIM/vimrc.tiny"
+#endif
 #ifndef SYS_VIMRC_FILE
 # define SYS_VIMRC_FILE "$VIM/vimrc"
 #endif
diff --git a/src/structs.h b/src/structs.h
index 5d0541b..d2ca1fc 100644
--- a/src/structs.h
+++ b/src/structs.h
@@ -3472,6 +3472,9 @@ typedef struct
 #ifdef FEAT_DIFF
     int		diff_mode;		/* start with 'diff' set */
 #endif
+#ifdef SYS_TINYRC_FILE
+    int		vi_mode;		/* started as "vi" */
+#endif
 } mparm_T;
 
 /*
